src/parser.ts

```diff
@@ -0,0 +1,759 @@
+// ----- Imports ----- //
+
+import type { Option, Result } from '@guardian/types';
+import {
+	err,
+	ok,
+	resultAndThen,
+	ResultKind,
+	resultMap,
+	toOption,
+} from '@guardian/types';
+
+// ----- Types ----- //
+
+interface Parser<A> {
+	run: (a: unknown) => Result<string, A>;
+}
+
+// ----- Core Functions ----- //
+
+const parser = <A>(f: (a: unknown) => Result<string, A>): Parser<A> => ({
+	run: f,
+});
+
+/**
+ * Ignores whatever the input is and instead provides a successful parsing
+ * result containing `a`.
+ *
+ * @param a The value to be contained in the successful parsing result
+ * @returns {Parser<A>} A `Parser` of `A`
+ * @example
+ * const fooParser = succeed('foo'); // Parser<string>
+ *
+ * const result = parse(fooParser)(42); // Ok<string>, with value 'foo'
+ */
+const succeed = <A>(a: A): Parser<A> => parser((_) => ok(a));
+
+/**
+ * Ignores whatever the input is and instead provides a failed parsing
+ * result containing the error string `e`.
+ *
+ * @param e The error string to be provided as the parsing error
+ * @returns {Parser<A>} A `Parser` of `A`
+ * @example
+ * const fooParser = fail('Uh oh!'); // Parser<A>
+ *
+ * const result = parse(fooParser)(42); // Err<string>, with value 'Uh oh!'
+ */
+const fail = <A>(e: string): Parser<A> => parser((_) => err(e));
+
+const isObject = (a: unknown): a is Record<string, unknown> =>
+	typeof a === 'object' && a !== null;
+
+/**
+ * Run a parser over a given input that's currently `unknown`, and attempt
+ * to parse that input into a value of type `A`. If the parsing was successful
+ * the result will be an `Ok<A>`. If it failed the result will be an
+ * `Err<string>`.
+ *
+ * @param pa The parser to run over the input
+ * @param a The `unknown` input to be parsed into type-safe value of type `A`
+ * @returns {Result<string, A>} A parsing result, wrapped in a `Result` type.
+ * @example
+ * const json: unknown = JSON.parse('{ "foo": 42 }');
+ *
+ * const parserA = fieldParser('foo', numberParser); // Parser<number>
+ * const resultA = parse(parserA)(json); // Ok<number>, with value 42
+ *
+ * const parserB = fieldParser('bar', numberParser); // Parser<number>
+ * const resultB = parse(parserB)(json); // Err<string>, with 'missing field' err
+ */
+const parse = <A>(pa: Parser<A>) => (a: unknown): Result<string, A> =>
+	pa.run(a);
+
+// ----- Basic Parsers ----- //
+
+/**
+ * Parses a value into a `string`.
+ */
+const stringParser: Parser<string> = parser((a) =>
+	typeof a === 'string'
+		? ok(a)
+		: err(`Unable to parse ${String(a)} as a string`),
+);
+
+/**
+ * Parses a value into a `number`.
+ */
+const numberParser: Parser<number> = parser((a) =>
+	typeof a === 'number' && !isNaN(a)
```

Comment:
Yeah, I guess this could go either way. On balance, I think in most cases I've come across it's more useful to treat something that's not a number as a parse failure than a `NaN`. For example, in:

```ts
const a = parseInt('foo');
```

`a` is a `NaN`. It will trickle through your code as such, converting the result of all mathematical operations like `a * 5` into `NaN`s, until it eventually makes its way into a piece of UI: `"NaN comments on this article"`. I'd prefer to catch that early.

Reply:
👍